File: app/schemas/shortlist.py
"""
schemas/shortlist.py – Pydantic schemas for the shortlist / accept flow.

States:
  shortlisted → accepted | rejected
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.shortlist import ShortlistStatus
from app.schemas.profile import ProfileRead


class ShortlistCreate(BaseModel):
    """POST /shortlists – express interest in a profile."""

    to_profile_id: uuid.UUID
    note: str | None = Field(None, max_length=500)


class ShortlistStatusUpdate(BaseModel):
    """PATCH /shortlists/{id} – accept or reject an expression of interest."""

    status: ShortlistStatus


class ShortlistRead(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    from_profile_id: uuid.UUID
    to_profile_id: uuid.UUID
    status: ShortlistStatus
    note: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ShortlistList(BaseModel):
    items: list[ShortlistRead]
    total: int


class ProfileSummary(BaseModel):
    """Lightweight profile summary embedded in admin pair view."""
    id: uuid.UUID
    full_name: str | None = None
    gender: str | None = None
    date_of_birth: str | None = None
    city: str | None = None
    state: str | None = None

    model_config = {"from_attributes": True}


class ShortlistPairRead(BaseModel):
    """Admin view of a shortlist pair with embedded profile summaries."""
    id: uuid.UUID
    from_profile: ProfileSummary
    to_profile: ProfileSummary
    status: ShortlistStatus
    note: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ShortlistPairList(BaseModel):
    items: list[ShortlistPairRead]
    total: int
    page: int
    size: int
    pages: int


class InterestRead(BaseModel):
    """A shortlist entry enriched with the counterpart's full profile.

    Used by:
      GET /shortlists/sent-interests     → profile = to_profile (recipient)
      GET /shortlists/received-interests → profile = from_profile (sender)
    """

    shortlist_id: uuid.UUID
    status: ShortlistStatus
    note: str | None
    created_at: datetime
    profile: ProfileRead


class InterestList(BaseModel):
    items: list[InterestRead]
    total: int
    page: int
    size: int
    pages: int

File: frontend/src/api/shortlist.ts
import client from './client'
import type { ShortlistEntry, ShortlistList, ShortlistStatus, PaginatedResponse, Profile, InterestList } from '@/types'

export const shortlistApi = {
  /** Express interest in a profile */
  create(to_profile_id: string, note?: string): Promise<ShortlistEntry> {
    return client.post('/shortlists/', { to_profile_id, note: note ?? null }).then((r) => r.data)
  },

  /** Fetch all shortlist entries sent by the current user (up to limit) */
  sent(limit = 500): Promise<ShortlistList> {
    return client.get('/shortlists/sent', { params: { skip: 0, limit } }).then((r) => r.data)
  },

  /** Withdraw interest by shortlist entry ID */
  delete(id: string): Promise<void> {
    return client.delete(`/shortlists/${id}`).then(() => undefined)
  },

  /** Remove a shortlist entry by the target profile's ID */
  deleteByProfile(toProfileId: string): Promise<void> {
    return client.delete(`/shortlists/by-profile/${toProfileId}`).then(() => undefined)
  },

  /** Browse shortlisted profiles – same paginated response shape as GET /profiles/ */
  shortlistedProfiles(page = 1, size = 12): Promise<PaginatedResponse<Profile>> {
    return client
      .get('/shortlists/shortlisted-profiles', { params: { page, size } })
      .then((r) => r.data)
  },

  /**
   * Sent interests with full recipient profile + status.
   * Used in "My Interests → Sent" tab.
   */
  sentInterests(page = 1, size = 20): Promise<InterestList> {
    return client
      .get('/shortlists/sent-interests', { params: { page, size } })
      .then((r) => r.data)
  },

  /**
   * Received interests with full sender profile + status.
   * Used in "My Interests → Received" tab.
   */
  receivedInterests(page = 1, size = 20, status?: string): Promise<InterestList> {
    return client
      .get('/shortlists/received-interests', {
        params: { page, size, ...(status ? { status } : {}) },
      })
      .then((r) => r.data)
  },

  /**
   * Accept or reject a received interest.
   * Only the recipient can call this.
   */
  respond(shortlistId: string, status: 'accepted' | 'rejected'): Promise<ShortlistEntry> {
    return client
      .patch(`/shortlists/${shortlistId}`, { status })
      .then((r) => r.data)
  },
}
